src/data/nat/digits.lean

```diff
@@ -0,0 +1,292 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import data.nat.basic
+import algebra.euclidean_domain
+import tactic.ring
+import tactic.interval_cases
+
+/-!
+# Digits of a natural number
+
+This provides a basic API for extracting the digits of a natural number in a given base,
+and reconstructing numbers from their digits.
+
+We also prove some divisibility tests based on digits, in particular completing
+Theorem #85 from https://www.cs.ru.nl/~freek/100/.
+-/
+
+
+/-- (Impl.) An auxiliary definition for `digits`, to help get the desired definitional unfolding. -/
+def digits_aux_0 : ℕ → list ℕ
+| 0 := []
+| (n+1) := [n+1]
+
+/-- (Impl.) An auxiliary definition for `digits`, to help get the desired definitional unfolding. -/
+def digits_aux_1 (n : ℕ) : list ℕ := list.repeat 1 n
+
+/-- (Impl.) An auxiliary definition for `digits`, to help get the desired definitional unfolding. -/
+def digits_aux (b : ℕ) (h : 2 ≤ b) : ℕ → list ℕ
+| 0 := []
+| (n+1) :=
+  have (n+1)/b < n+1 := nat.div_lt_self (nat.succ_pos _) h,
+  (n+1) % b :: digits_aux ((n+1)/b)
+
+@[simp] lemma digits_aux_zero (b : ℕ) (h : 2 ≤ b) : digits_aux b h 0 = [] := rfl
+
+lemma digits_aux_def (b : ℕ) (h : 2 ≤ b) (n : ℕ) (w : 0 < n) :
+  digits_aux b h n = n % b :: digits_aux b h (n/b) :=
+begin
+  cases n,
+  { cases w, },
+  { rw [digits_aux], }
+end
+
+/--
+`digits b n` gives the digits, in little-endian order,
+of a natural number `n` in a specified base `b`.
+
+In any base, we have `of_digits b L = L.foldr (λ x y, x + b * y) 0`.
+* For any `2 ≤ b`, we have `l < b` for any `l ∈ digits b n`,
+  and this uniquely specifies the behaviour of `digits b`.
+* For `b = 1`, we define `digits 1 n = list.repeat 1 n`.
+* For `b = 0`, we define `digits 0 n = [n]`, except `digits 0 0 = []`.
+-/
```

Comment:
Maybe it's worth mentioning here that this is the same as [`nat.to_digits`](https://github.com/leanprover-community/lean/blob/fb3f3e8/library/init/data/repr.lean#L103) as long as `b >= 2` (I think).

Reply:
Oh, I failed to even find that....

Reply:
I've added a note to the effect that this definition is not quite the same as that one!